generators/commons/src/Writer.ts

```diff
@@ -28,20 +28,38 @@ export class Writer {
         const indentedText = indent + text.replace(/\n/g, `\n${indent}`);
 
         this.buffer += indentedText;
-
         this.hasWrittenAnything = true;
     }
 
     public newLine(): void {
         this.buffer += "\n";
     }
 
-    public openIndent(): void {
-        this.indentLevel++;
+    public openBlock(
+        titles: (string | undefined)[],
+        openingCharacter: string | undefined = "{",
+        callback: () => void,
+        closingCharacter: string | undefined = "}"
+    ): void {
```

Comment:
nit: Something we've become fans of is in-lined request objects. That way, each positional parameter is named which reduces the amount of time you need to jump back on forth between a signature to know that you've ordered it correctly. This is particularly relevant for functions that accept two parameters of the same type.

I'd refactor this like so:
```suggestion
    public openBlock({
        titles,
        openingCharacter,
        callback,
        closingCharacter,
    }: {
        titles: (string | undefined)[];
        openingCharacter: string | undefined = "{";
        callback: () => void;
        closingCharacter: string | undefined = "}";
    }): void {
```

This will require updating every call-site, too.

Reply:
Now that I've reviewed all the call-sites more, is this function even worth it? It's trying to cram in a lot of functionality into one place. It's a generic block writer for classes, function definitions, and structs all in one place (presumably for multiple languages).

Should we instead just explicitly write out each definition explicitly so it's easier to read? We could still capture an indentation helper with a separate `writer.indent` method (or something similar). I imagine you could include something like the following for the `class` definition you have below:

```typescript
    public write(writer: Writer): void {

        // e.g. public class ClassName {
        if (this.accessLevel != null) {
          writer.write(this.accessLevel);
          writer.write(" ");
        }
        writer.writeLine(`class ${this.name} {`);
 
        writer.indent(() => {
          // TODO: Add fields
          // Add functions
          ...
        });

        writer.writeLine("}")
    }
```

What are your thoughts? Not married to anything here, but trying to consider the simplest / easiest to read solution.